lib/tree.rb

```diff
@@ -16,48 +16,200 @@ def initialize
+  # Space Complexity: O(1) the amount of data stored is constant, though the space required in memory for the stack trace will be O(logn)
   def find(key)
-    raise NotImplementedError
+    if @root.nil?
+      return nil
+    else
+      return find_helper(@root, key)
+    end
+  end
+
+  def find_helper(current, key)
+    if current.key == key
+      return current.value
+    end
+
+    if key < current.key
+      find_helper(current.left, key)
+    else
+      find_helper(current.right, key)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n), because the algorithm will have to visit every node on the tree to add them to the nodes_array list
+  # Space Complexity: O(n) because the size of the nodes_array list will vary linearly with the number of nodes in the tree.
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])  
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(current, nodes_array)
+    unless current.nil?
+      inorder_helper(current.left, nodes_array)
+      nodes_array << { key: current.key, value: current.value }
+      inorder_helper(current.right, nodes_array)
+    end
+
+    return nodes_array
+  end
+
+  # Time Complexity: O(n), because the algorithm will have to visit every node on the tree to add them to the nodes_array list
+  # Space Complexity: O(n) because the size of the nodes_array list will vary linearly with the number of nodes in the tree.
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
+  end
+
+  def preorder_helper(current, nodes_array)
+    unless current.nil?
+      nodes_array << { key: current.key, value: current.value }
+      preorder_helper(current.left, nodes_array)
+      preorder_helper(current.right, nodes_array)
+    end
+
+    return nodes_array
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n), because the algorithm will have to visit every node on the tree to add them to the nodes_array list
+  # Space Complexity: O(n) because the size of the nodes_array list will vary linearly with the number of nodes in the tree. 
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
+  end
+
+  def postorder_helper(current, nodes_array)
+    unless current.nil?
+      postorder_helper(current.left, nodes_array)
+      postorder_helper(current.right, nodes_array)
+      nodes_array << { key: current.key, value: current.value }
+    end
+
+    return nodes_array
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n), because the algorithm will have to visit every node on the tree in order to determine the height of the tree
+  # Space Complexity: O(1) because the amount of data stored is constant, though the space required in memory for the stack trace will be O(n)
   def height
```

Comment:
⚠️ Since this method is recursive, you have to account for the call stack in terms of space complexity.